File: typescript/factory_method/id-card-factory.ts
import { Factory } from './factory';
import { Product } from './product';
import { IdCard } from './id-card';

export class IdCardFactory extends Factory {
  private _owners: string[] = [];

  createProduct(owner: string): Product {
    return new IdCard(owner);
  }

  registerProduct(product: Product): void {
    const card = product as IdCard;
    this._owners.push(card.owner);
  }

  get owners(): string[] {
    return this._owners;
  }
}

File: kotlin/factory_method/src/IdCardFactory.kt
package factoryMethod

class IdCardFactory : Factory() {
    val owners = arrayListOf<String>()

    override fun createProduct(owner: String): Product {
        return IdCard(owner)
    }

    override fun registerProduct(product: Product) {
        val card = product as IdCard
        this.owners.add(card.owner)
    }
}